File: shared/Honeycomb/Models/HoneycombRuleSelectionEngine.swift
import Foundation

// Shared Auto/Pick/Ban logic for the Rules screen's unified rule list, used by both
// mac's HoneycombRulesView and iOS's merged rules sheet so the selection/ban validation
// (max-4 cap, exclusive rule pairs, "can't ban everything" guard) lives in one place
// instead of being copy-pasted per platform.
public enum HoneycombRuleState {
    case auto
    case picked
    case banned
}

// "Normal Mode" isn't a HoneycombRule case — picking it means `forceNormalMode = true`
// (lock the match to zero rules) and banning it means `bannedRules.contains("Normal Mode")`
// (keep roulette from ever landing on a zero-rule match). Modeling it alongside the real
// rules here is what lets the unified list treat every row the same way.
public enum HoneycombRuleRowID: Hashable {
    case rule(HoneycombRule)
    case normalMode

    public static let banListOrder: [HoneycombRuleRowID] = [.normalMode] + HoneycombRule.allCases.map { .rule($0) }
    public static let pickableOrder: [HoneycombRuleRowID] = [.normalMode] + HoneycombRule.allCases.map { .rule($0) }

    public var isPickable: Bool { true }

    var banName: String {
        switch self {
        case .normalMode: return "Normal Mode"
        case .rule(let r): return r.rawValue
        }
    }
}

public enum HoneycombRuleSelection {
    public static func state(
        of id: HoneycombRuleRowID,
        selectedRules: Set<HoneycombRule>,
        forceNormalMode: Bool,
        bannedRules: Set<String>
    ) -> HoneycombRuleState {
        if bannedRules.contains(id.banName) { return .banned }
        switch id {
        case .normalMode: return forceNormalMode ? .picked : .auto
        case .rule(let r): return selectedRules.contains(r) ? .picked : .auto
        }
    }

    // Returns true if the ban was blocked by the "can't ban everything" guard (the
    // caller shows the silly-bee warning in that case); false otherwise.
    @discardableResult
    public static func setState(
        _ newState: HoneycombRuleState,
        for id: HoneycombRuleRowID,
        selectedRules: inout Set<HoneycombRule>,
        forceNormalMode: inout Bool,
        bannedRules: inout Set<String>
    ) -> Bool {
        switch newState {
        case .auto:
            bannedRules.remove(id.banName)
            if case .normalMode = id { forceNormalMode = false }
            if case .rule(let r) = id { selectedRules.remove(r) }
            return false

        case .picked:
            guard id.isPickable else { return false }
            bannedRules.remove(id.banName)
            switch id {
            case .normalMode:
                forceNormalMode = true
                selectedRules.removeAll()
            case .rule(let r):
                var updated = selectedRules
                // Remove the exclusive partner (if any) BEFORE the cap check — selecting
                // a rule whose partner is already selected is a net-zero swap, not an
                // addition, so it must never be blocked just because the cap is full.
                if r == .ascension { updated.remove(.descension) }
                if r == .descension { updated.remove(.ascension) }
                if r == .order { updated.remove(.chaos) }
                if r == .chaos { updated.remove(.order) }
                if r == .allOpen { updated.remove(.threeOpen) }
                if r == .threeOpen { updated.remove(.allOpen) }
                // Bomb Shelter's hidden card doesn't work when All Open/Three Open
                // reveals every card anyway.
                if r == .allOpen || r == .threeOpen { updated.remove(.bombShelter) }
                if r == .bombShelter { updated.remove(.allOpen); updated.remove(.threeOpen) }

                if updated.count < 4 {
                    updated.insert(r)
                    selectedRules = updated
                    forceNormalMode = false
                }
            }
            return false

        case .banned:
            // "Silly bee" guard: never allow banning the last remaining unbanned item.
            if bannedRules.count >= HoneycombRuleRowID.banListOrder.count - 1 && !bannedRules.contains(id.banName) {
                return true
            }
            if case .normalMode = id { forceNormalMode = false }
            if case .rule(let r) = id { selectedRules.remove(r) }
            bannedRules.insert(id.banName)
            return false
        }
    }
}

File: android/app/src/main/java/com/leah/honeycomb/honeycomb/HoneycombRuleSelectionEngine.kt
package com.leah.honeycomb.honeycomb

enum class HoneycombRuleState {
    Auto, Picked, Banned
}

sealed class HoneycombRuleRowID {
    data class Rule(val rule: HoneycombRule) : HoneycombRuleRowID()
    object NormalMode : HoneycombRuleRowID()

    val isPickable: Boolean get() = true

    val banName: String
        get() = when (this) {
            is NormalMode -> "Normal Mode"
            is Rule -> rule.name // Use enum name for serialization consistency
        }

    companion object {
        val allCases: List<HoneycombRuleRowID> by lazy {
            listOf(NormalMode) + HoneycombRule.values().map { Rule(it) }
        }
    }
}

object HoneycombRuleSelection {
    fun state(
        id: HoneycombRuleRowID,
        selectedRules: Set<HoneycombRule>,
        forceNormalMode: Boolean,
        bannedRules: Set<String>
    ): HoneycombRuleState {
        if (bannedRules.contains(id.banName)) return HoneycombRuleState.Banned
        return when (id) {
            is HoneycombRuleRowID.NormalMode -> if (forceNormalMode) HoneycombRuleState.Picked else HoneycombRuleState.Auto
            is HoneycombRuleRowID.Rule -> if (selectedRules.contains(id.rule)) HoneycombRuleState.Picked else HoneycombRuleState.Auto
        }
    }

    // Returns true if the ban was blocked by the "can't ban everything" guard.
    fun setState(
        newState: HoneycombRuleState,
        id: HoneycombRuleRowID,
        selectedRules: MutableSet<HoneycombRule>,
        forceNormalMode: MutableList<Boolean>, // Using list size 1 to mutate reference
        bannedRules: MutableSet<String>
    ): Boolean {
        when (newState) {
            HoneycombRuleState.Auto -> {
                bannedRules.remove(id.banName)
                if (id is HoneycombRuleRowID.NormalMode) forceNormalMode[0] = false
                if (id is HoneycombRuleRowID.Rule) selectedRules.remove(id.rule)
                return false
            }
            HoneycombRuleState.Picked -> {
                if (!id.isPickable) return false
                bannedRules.remove(id.banName)
                when (id) {
                    is HoneycombRuleRowID.NormalMode -> {
                        forceNormalMode[0] = true
                        selectedRules.clear()
                    }
                    is HoneycombRuleRowID.Rule -> {
                        val r = id.rule
                        if (r == HoneycombRule.Ascension) selectedRules.remove(HoneycombRule.Descension)
                        if (r == HoneycombRule.Descension) selectedRules.remove(HoneycombRule.Ascension)
                        if (r == HoneycombRule.Order) selectedRules.remove(HoneycombRule.Chaos)
                        if (r == HoneycombRule.Chaos) selectedRules.remove(HoneycombRule.Order)
                        if (r == HoneycombRule.AllOpen) selectedRules.remove(HoneycombRule.ThreeOpen)
                        if (r == HoneycombRule.ThreeOpen) selectedRules.remove(HoneycombRule.AllOpen)
                        if (r == HoneycombRule.AllOpen || r == HoneycombRule.ThreeOpen) selectedRules.remove(HoneycombRule.BombShelter)
                        if (r == HoneycombRule.BombShelter) {
                            selectedRules.remove(HoneycombRule.AllOpen)
                            selectedRules.remove(HoneycombRule.ThreeOpen)
                        }

                        if (selectedRules.size < 4) {
                            selectedRules.add(r)
                            forceNormalMode[0] = false
                        }
                    }
                }
                return false
            }
            HoneycombRuleState.Banned -> {
                if (bannedRules.size >= HoneycombRuleRowID.allCases.size - 1 && !bannedRules.contains(id.banName)) {
                    return true
                }
                if (id is HoneycombRuleRowID.NormalMode) forceNormalMode[0] = false
                if (id is HoneycombRuleRowID.Rule) selectedRules.remove(id.rule)
                bannedRules.add(id.banName)
                return false
            }
        }
    }
}
